src/wallet/multi_tracker.rs

```diff
@@ -0,0 +1,405 @@
+    fn new_address(&self, desc: &ExtendedDescriptor) -> Result<AddressInfo, Error>;
+}
+
+/// Represents a collection of owned and spendable `UTXO`s, `ExtendedDescriptor`s and associated
+/// `Transaction`s.
+pub trait MultiTracker: MultiTrackerInner {
+    /// Obtains parent [`ExtendedDescriptor`] and child index of the provided `ScriptPubKey`.
+    /// Note that if the script is not owned, None shoud be returned.
+    fn get_path_of_script_pubkey(
+        &self,
+        script: &Script,
+    ) -> Result<Option<(DescriptorItem, u32)>, Error> {
+        let descriptors = self.iter_descriptors()?.collect::<Vec<_>>();
+
+        let secp = SecpCtx::new();
+
+        for index in 0..2100_u32 {
+            for item in &descriptors {
+                let desc = item.descriptor.clone();
+                let derived_script = desc.as_derived(index, &secp).script_pubkey();
+                if script == &derived_script {
+                    return Ok(Some((item.clone(), index)));
+                }
+            }
+        }
+
+        Ok(None)
+    }
+
+    /// Obtain a local UTXO given the provided outpoint.
+    /// The default implementation is very inefficient and should be overloaded.
+    fn get_utxo(&self, outpoint: &OutPoint) -> Result<Option<LocalUtxo>, Error> {
+        Ok(self.iter_unspent()?.find_map(|(utxo, _)| {
+            if &utxo.outpoint == outpoint {
+                Some(utxo)
+            } else {
+                None
+            }
+        }))
+    }
+
+    /// Obtain output of the provided outpoint.
+    /// Output may not be owned by us, just be part of a transaction that we are keeping track of.
+    fn get_output(&self, outpoint: &OutPoint) -> Result<Option<TxOut>, Error> {
+        self.get_tx(&outpoint.txid)?.map_or_else(
+            || Err(Error::InvalidOutpoint(*outpoint)),
+            |tx_item| {
+                Ok(tx_item
+                    .raw
+                    .output
+                    .get(outpoint.vout as usize)
+                    .map(Clone::clone))
+            },
+        )
+    }
+
+    /// Returns whether given script is owned by us, or not.
+    fn is_mine(&self, script: &Script) -> Result<bool, Error> {
+        self.get_path_of_script_pubkey(script)
+            .map(|path| path.is_some())
+    }
+
+    /// Obtains a new change address without a descriptor provided explicitly
+    #[deprecated]
+    fn new_auto_address(&self) -> Result<AddressInfo, Error> {
+        let (mut internal, mut external): (Vec<_>, Vec<_>) = self
+            .iter_descriptors()?
+            .partition(|item| item.keychain == KeychainKind::Internal);
+
+        internal.append(&mut external);
+
+        let item = internal.iter().chain(external.iter()).next().unwrap();
+
+        #[allow(deprecated)]
+        self.new_address(&item.descriptor)
+    }
+}
+
+/// Implements [`MultiTracker`] with one external descriptor and an optional internal descriptor.
+// #[derive(Clone)]
+pub struct LegacyTracker<'a, D> {
```

Comment:
@afilini It is just to avoid copying data in memory. The assumption is that one would only use `LegacyTracker` with the existence of `Wallet` (which is a good assumption imo). Since, the whole point of `LegacyTracker` is to get `Wallet` to "implement" `MultiTracker`, since transaction-building and signing logic is to be wrapped around `MultiTracker`. Right now this tx logic uses static functions of `Wallet` (which is confusing), but will avoid annoying git as much as possible. Those static functions, responsible for tx logic, should eventually be moved out of `Wallet` when the time is right.